test/expect-tests/fiber/fiber_tests.ml

```diff
@@ -642,48 +642,59 @@ let%expect_test "Stream.parallel_iter doesn't leak" =
      This test is carefully crafted to avoid creating new live words as we
      iterate through the stream. As a result, the only new live words that can
      appear are because of the iteration function. *)
-  let test ~iter_function ~check =
-    let stream n =
-      let n = ref n in
+  let test n ~iter_function =
+    let finish_stream = Fiber.Ivar.create () in
+    let stream =
+      let count = ref n in
       Fiber.Stream.In.create (fun () ->
-          if !n = 0 then
-            Fiber.return None
-          else (
-            decr n;
+          if !count > 0 then (
+            decr count;
             Fiber.return (Some ())
-          ))
+          ) else
+            let* () = Fiber.Ivar.read finish_stream in
+            Fiber.return None)
     in
-    (* We use [-1] as a [None] value to avoid going from [None] to [Some _],
-       which would case the number of live words to change *)
-    let prev = ref (-1) in
-    let ok = ref true in
-    let f () =
+    let awaiting = ref n in
+    let iter_await = Fiber.Ivar.create () in
+    let record () =
       Gc.full_major ();
       let curr = (Gc.stat ()).live_words in
-      if !prev >= 0 then
-        if not (check ~prev:!prev ~curr) then (
-          Printf.printf
-            "[FAIL] live words not changing as expected: prev=%d, curr=%d\n"
-            !prev curr;
-          ok := false
-        );
-      prev := curr;
-      Fiber.return ()
+      curr
+    in
+    let f () =
+      decr awaiting;
+      if !awaiting = 0 then
+        Fiber.Ivar.fill iter_await ()
+      else
+        Fiber.return ()
     in
-    Scheduler.run (iter_function (stream 100) ~f);
-    if !ok then print_string "PASS"
+    Scheduler.run
+      (let+ prev, curr =
+         Fiber.fork_and_join
+           (fun () ->
+             let prev = record () in
+             let+ () = iter_function stream ~f in
+             prev)
+           (fun () ->
+             let* () = Fiber.Ivar.read iter_await in
+             let curr = record () in
+             let+ () = Fiber.Ivar.fill finish_stream () in
+             curr)
+       in
+       curr - prev)
   in
 
+  let noise = test 1 ~iter_function:Fiber.Stream.In.parallel_iter in
+
   (* Check that the number of live words keeps on increasing because we are
      leaking memory: *)
-  test ~iter_function:naive_stream_parallel_iter ~check:(fun ~prev ~curr ->
-      prev < curr);
-  [%expect {| PASS |}];
+  let delta = test 1_000 ~iter_function:naive_stream_parallel_iter in
+  assert (delta > noise);
```

Comment:
This could be a false positive. For instance, maybe `test 1 ~iter_function:native_stream_parallel_iter` is already higher than `test 1 ~iter_function:Fiber.Stream.In.parallel_iter` for some reason.

I think we should run `test n ~iter_function:native_stream_parallem_iter` for increasing `n` values and check that the results is increasing with `n`, and then do the same with `Fiber.Stream.In.parallel_iter` and check that the result is constant.